### PR TITLE
initial support for docker

### PR DESCRIPTION
**Pull request by jsuchome**

### CODE REVIEW

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -2187,6 +2191,13 @@ function oncontroller_testsetup()
         fi
     fi
 
+    if [[ -n "$wantdocker" ]] ; then
```
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: This is making a huge function even longer.  It would be better if it called a separate function.

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -2431,6 +2442,14 @@ EOF
         fi
     fi
 
+    if [ -n "$wantdocker" ]; then
```
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: Same comment here.  Alarm bells should be ringing every time you have to add code to a long function (or method).

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -2303,6 +2315,18 @@ function oncontroller()
     return $?
 }
 
+function get_docker_image()
+{
+    if [ -n "$want_docker" ]; then
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: IMHO, this if should not be inside this function. The function should just do what its name suggests, and not depend on another flag.

File: scripts/qa_crowbarsetup.sh [hunk 4]
```diff
@@ -2431,6 +2455,7 @@ EOF
         fi
     fi
 
+    get_docker_image
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: The if clause should rather be here.

File: scripts/qa_crowbarsetup.sh [hunk 5]
```diff
@@ -2170,27 +2174,35 @@ function oncontroller_testsetup()
...
         elif [[ -n "$wantxenpv" ]] ; then
-            glance --insecure image-create --name=SP3-64 --is-public=True --disk-format=qcow2 --container-format=bare --property hypervisor_type=xen --property vm_mode=xen --copy-from http://clouddata.cloud.suse.de/images/jeos-64-pv.qcow2 | tee glance.out
+            glance --insecure image-create --name=$image_name --is-public=True --disk-format=qcow2 --container-format=bare --property hypervisor_type=xen --property vm_mode=xen --copy-from http://clouddata.cloud.suse.de/images/jeos-64-pv.qcow2 | tee glance.out
         else
-            glance image-create --name=SP3-64 --is-public=True --property vm_mode=hvm --disk-format=qcow2 --container-format=bare --copy-from http://clouddata.cloud.suse.de/images/SP3-64up.qcow2 | tee glance.out
+            glance image-create --name=$image_name --is-public=True --property vm_mode=hvm --disk-format=qcow2 --container-format=bare --copy-from http://clouddata.cloud.suse.de/images/SP3-64up.qcow2 | tee glance.out
         fi
     fi
 
+    if [[ -n "$want_docker" ]] ; then
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: just nitpicking:
with double brackets use `if [[ $want_docker ]]`
or with single brackets use `if [ -n "$want_docker" ]`

**Comment by jdsn** [comment]
:+1:

**Comment by bmwiedemann** [comment]
I dont like how the image_name variable reduces copy-pastability. Maybe we can change it to "testimage" to not confuse people.

**Comment by jsuchome** [comment]
@bmwiedemann I do not understand. What's wrong with variable name?

**Comment by bmwiedemann** [comment]
The variable means that when I want to copy-paste such a line into a live cloud, I first need another line to fill the variable, thus slightly reducing the read-it-as-documentation-value of the code

**Comment by jsuchome** [comment]
@bmwiedemann the problem with docker image is, its glance image name has to be the same as docker  image. Hence, no universal 'testimage' here and the need for variable

File: scripts/qa_crowbarsetup.sh [hunk 6]
```diff
@@ -2303,6 +2315,16 @@ function oncontroller()
     return $?
 }
 
+function get_docker_image()
+{
+    # docker can only be run from SLE12 node
+    local dockernode=$(knife search node "roles:nova-multi-compute-docker" -a name | grep ^name: | cut -d : -f 2 | tail -n 1 | sed 's/\s//g')
+    ssh $dockernode "docker pull cirros && docker save -o cirros-docker.tar cirros"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: So I wonder: do we really require the `docker pull`, or if we save the image on clouddata and upload it in glance from there, it'll work?

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: That seems correct. In such case, we could drop this get_docker_image function...

**Comment by vuntz** [comment]
Not tested, but looks good to me.

File: scripts/mkcloud [hunk 7]
```diff
@@ -1277,6 +1277,8 @@ Optional
         if there is a SLE12 node, deploy neutron-network role into the SLE12 node
     want_dvr=1 (default='')
         if DVR support should be turned on by neutron barclamp. Only works with openvswitch and vxlan.
+    want_docker=1 (default='')
+        Test docker image (uploading to glance, creating instance based on it etc.)
```
Comment (MaximilianMeister) [inline scripts/mkcloud]: i would just say deploy a docker compute node or sth. because testing it is optional and dependent on the `testsetup` step, isn't it?
maybe also mention that it only works for a sle12 node, or add a sanity check that won't let you `want_docker` without setting `want_sles12`

File: scripts/qa_crowbarsetup.sh [hunk 8]
```diff
@@ -2170,27 +2174,34 @@ function oncontroller_testsetup()
     nova list
     glance image-list
 
-    if glance image-list | grep -q SP3-64 ; then
-        glance image-show SP3-64 | tee glance.out
+    image_name="SP3-64"
```
Comment (MaximilianMeister) [inline scripts/qa_crowbarsetup.sh]: please make this variable `local` and i personally would put this into a separate commit, as it does not really relate to the docker enablement

**Comment by MaximilianMeister** [comment]
I tested it with `want_sles12` and `want_docker` and it failed with 
```
Failed to edit: default : Errors in data
Failed to validate proposal: Node d52-54-02-77-77-02.v7.cloud.suse.de has been assigned to a nova-multi-compute role more than once
Error: Error: 'crowbar nova proposal --file=/root/nova.default.proposal edit default' failed with exit code: 1
```

**Comment by MaximilianMeister** [comment]
@jsuchome I think you should take out the docker compute node from the `$novanodes` somehow. the same node will be used for `nova-multi-compute-kvm` as well

**Comment by jsuchome** [comment]
@MaximilianMeister thanks for testing, could you try again?

**Comment by jsuchome** [comment]
Hm, no, removing from nodescompute is not right, it would not get e.g. ceilometer role...

**Comment by jsuchome** [comment]
Well, mkcloud kind of works for me now. It boots docker instance, but fails bacause of tempest failures (http://pastebin.suse.de/13658/src).. not sure if we can do more here. @vuntz ?

**Comment by vuntz** [comment]
@jsuchome I guess the issue is that tempest is not configured to use the right image for booting instances. So we'd need to change that in the tempest barclamp, but it's unclear how to best do it in a way that is not too complicated :/

**Comment by vuntz** [comment]
https://github.com/crowbar/barclamp-tempest/pull/140 should help with the tempest bits (although there are still some failures that are expected; not sure how to "ignore" them).

File: scripts/qa_crowbarsetup.sh [hunk 9]
```diff
@@ -1711,6 +1713,10 @@ function custom_configuration()
             proposal_set_value nova default "['attributes']['nova']['use_migration']" "true"
             [[ "$libvirt_type" = xen ]] && sed -i -e "s/nova-multi-compute-$libvirt_type/nova-multi-compute-xxx/g; s/nova-multi-compute-kvm/nova-multi-compute-$libvirt_type/g; s/nova-multi-compute-xxx/nova-multi-compute-kvm/g" $pfile
 
+            if [[ $all_with_ssl = 1 || $nova_with_ssl = 1 ]] ; then
+                enable_ssl_for_nova
+            fi
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: This sounds unrelated to this PR. Error during a rebase? Or github displaying weird bits?

File: scripts/qa_crowbarsetup.sh [hunk 10]
```diff
@@ -2287,6 +2309,10 @@ function oncontroller_testsetup()
         complain 96 "VM not accessible in reasonable time, exiting."
     fi
 
+    if [ "$image_name" == "cirros" ]; then
+        vmip="cirros@$vmip"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Would it make sense to have a variable for the ssh user, and set it at the same time as the `image_name` variable?

File: scripts/qa_crowbarsetup.sh (hunk 9 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Hm, looks like incorrectly handled conflict during rebase...

File: scripts/qa_crowbarsetup.sh [hunk 11]
```diff
@@ -2304,6 +2330,15 @@ function oncontroller_testsetup()
     if ! ssh $vmip curl www3.zq1.de/test ; then
         complain 95 could not reach internet
     fi
+
+    # skip volume tests for docker
+    if [ "$image_name" == "cirros" ]; then
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Shouldn't the test be `if [ -n "$want_docker" ] ; then` (for consistency)?

File: scripts/qa_crowbarsetup.sh [hunk 12]
```diff
@@ -2304,6 +2330,15 @@ function oncontroller_testsetup()
...
+        echo "RadosGW Tests: $radosgwret"
+        echo "Tempest: $tempestret"
+        echo "Volume in VM: skipped (not supported for docker)"
+        test $tempestret = 0 -a $radosgwret = 0 || exit 102
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: I'm a bit worried that we're not removing the floating IP here. Maybe we should instead do something like:

```
if [ -z "$want_docker" ] ; then
  # do volume tests
fi

# cleanup so that we can run testvm without leaking volumes, IPs etc
...
```

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: I didn't want to indent all that volume stuff, but you're probably right...

Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: @jsuchome there's a good chance I misunderstood your comment here, but just wanted say in case it's relevant: splitting stuff off into functions helps minimise indentation levels, and is also generally a huge boost to code legibility and maintainability.

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: I know, but the code is already there, the amount of added lines here is minimal

File: scripts/qa_crowbarsetup.sh [hunk 13]
```diff
@@ -2287,6 +2307,8 @@ function oncontroller_testsetup()
         complain 96 "VM not accessible in reasonable time, exiting."
     fi
 
+    vmip="$ssh_user@$vmip"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Nitpick: maybe use a new variable? `ssh_target`?

File: scripts/qa_crowbarsetup.sh [hunk 14]
```diff
@@ -2304,30 +2326,39 @@ function oncontroller_testsetup()
...
+        ssh $vmip "mount $device /mnt && grep -q $rand /mnt/test.txt" || volumeattachret=58
+        # cleanup so that we can run testvm without leaking volumes, IPs etc
+        nova remove-floating-ip "$instanceid" "$floatingip"
+        nova floating-ip-delete "$floatingip"
+        nova stop "$instanceid"
+        wait_for 100 1 "test \"x\$(nova show \"$instanceid\" | perl -ne 'm/ status [ |]*([a-zA-Z]+)/ && print \$1')\" == xSHUTOFF" "testvm to stop"
+
+        echo "Volume in VM: $volumecreateret & $volumeattachret"
+    else
+        echo "Volume in VM: tests skipped (not supported for docker)"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: I think I'd leave the `echo` where it was and instead do something like:
```shell
local volume_result=""

if [ -z "$want_docker" ] ; then
    local volumecreateret=0
    local volumeattachret=0
    ...
    volume_result="$volumecreateret & $volumeattachret"
else
    volume_result="tests skipped (not supported by docker backend)"
fi
...
echo "Volume in VM: $volume_result"
```

But just my personal taste, keep it the current way if you prefer the way it is.

File: scripts/qa_crowbarsetup.sh [hunk 15]
```diff
@@ -2287,11 +2307,13 @@ function oncontroller_testsetup()
         complain 96 "VM not accessible in reasonable time, exiting."
     fi
 
+    ssh_target="$ssh_user@$vmip"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Missing `local`?

File: scripts/qa_crowbarsetup.sh (hunk 14 above)
Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: There's a check for volumecreateret and volumeattachret later, that's why I moved them outside the block

File: scripts/qa_crowbarsetup.sh [hunk 16]
```diff
@@ -2301,33 +2323,44 @@ function oncontroller_testsetup()
...
+        ssh $ssh_target "mkfs.ext3 -F $device && mount $device /mnt && echo $rand > /mnt/test.txt && umount /mnt"
+        nova volume-detach "$instanceid" "$volumeid" ; sleep 10
+        nova volume-attach "$instanceid" "$volumeid" /dev/vdb ; sleep 10
+        ssh $ssh_target fdisk -l $device | grep 1073741824 || volumeattachret=57
+        ssh $ssh_target "mount $device /mnt && grep -q $rand /mnt/test.txt" || volumeattachret=58
+        # cleanup so that we can run testvm without leaking volumes, IPs etc
+        nova remove-floating-ip "$instanceid" "$floatingip"
+        nova floating-ip-delete "$floatingip"
+        nova stop "$instanceid"
+        wait_for 100 1 "test \"x\$(nova show \"$instanceid\" | perl -ne 'm/ status [ |]*([a-zA-Z]+)/ && print \$1')\" == xSHUTOFF" "testvm to stop"
```
Comment (vuntz) [inline scripts/qa_crowbarsetup.sh]: Sorry, didn't notice that earlier, but the last 5 lines should be out of the `if` block. These were actually the reason why I suggested to reorganize the code :-)

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Ah! Sure, I know, I just forgot again ... :-(

**Comment by vuntz** [comment]
+1

File: scripts/qa_crowbarsetup.sh [hunk 17]
```diff
@@ -1736,6 +1738,16 @@ function custom_configuration()
                 novanodes="[ ${novanodes%,} ]"
                 proposal_set_value nova default "['deployment']['nova']['elements']['nova-multi-compute-${libvirt_type}']" "$novanodes"
             fi
+
+            if [ -n "$want_sles12" ] && [ -n "$want_docker" ] ; then
+                proposal_set_value nova default "['deployment']['nova']['elements']['nova-multi-compute-docker']" "['$sle12node']"
+                # do not assign another compute role to this node
+                local novanodes
+                novanodes=`printf "%s\n" $novanodes | grep -iv $sle12node`
+                novanodes="[ ${novanodes} ]"
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: This is still not a valid ruby value.
It mus look like this:
```
[ "value1", "value2", "value3" ]
```
Your code creates this:
```
[ value1  value2  value3 ]
```

Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: See also:  https://github.com/SUSE-Cloud/automation/blob/master/scripts/qa_crowbarsetup.sh#L1526
for how to do it.

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: It's the same code as the one at lines 1735-1737

File: scripts/qa_crowbarsetup.sh [hunk 18]
```diff
@@ -2489,6 +2522,12 @@ EOF
         fi
     fi
 
+    # prepare docker image at sle12 compute node
+    if [ -n "$want_sles12" ] && [ -n "$want_docker" ] ; then
+        local sle12node=$(knife search node "target_platform:suse-12.0" -a name | grep ^name: | cut -d : -f 2 | tail -n 1 | sed 's/\s//g')
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: Duplicated code.
For such tasks I think it is worth addin a function. Then the code here gets more readable and the magic can be done in a separate function. Plus, it can be reused by other people and it can be tested individually.

File: scripts/qa_crowbarsetup.sh (hunk 17 above)
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: No it isn't. Your code does not add the commas and the quotes.

Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: You most likely meant something like this:
```
+
+            if [ -n "$want_sles12" ] && [ -n "$want_docker" ] ; then
+                proposal_set_value nova default "['deployment']['nova']['elements']['nova-multi-compute-docker']" "['$sle12node']"
+                # do not assign another compute role to this node
+                local novanodes
+                novanodes=`printf "\"%s\",\n" $nodescompute | grep -iv $sle12node`
+                novanodes="[ ${novanodes%,} ]"
+                novanodes="[ ${novanodes} ]"

```

Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: This is another example of the kind of crap we have to deal with as a result of using shell :-/  Proper languages help you get the job done rather than waste your time by introducing stupid quoting issues when dealing with simple data structures.

Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: Its not 'another' - its stil the same, and we know it. We are working on refactoring mkcloud/qa_crowbarsetup with the goal to split it into smaller pieces (in order to transform the smaller pieces easier into other languages). This is tricky and takes time. Until then, this is an exercise to learn to be patient  :)

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: @jdsn ah, yes, thanks! I just wonder, because I think it was working before... maybe I've done some other change meanwhile

Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: > Its not 'another'

I said "another" because it's the second one I encountered this morning ;-/

> We are working on refactoring mkcloud/qa_crowbarsetup with the goal to split it into smaller pieces (in order to transform the smaller pieces easier into other languages).

Yes, I think I was the one who suggested doing that ;-)

> This is tricky and takes time.

Of course.

> Until then, this is an exercise to learn to be patient :)

Sure, I'm waiting patiently :-)  But I also wanted to highlight a real-world example of shell causing issues, to make sure that people know there are valid reasons for my rants about shell ;-)

**Comment by jdsn** [comment]
:+1: lgtm now, but before we merge it, please give it another test run after all these changes

**Comment by jsuchome** [comment]
Reseting that nova-multi-compute-* is wrong, I haven't realized that $nodescompute is empty in non-cluster case...

**Comment by jsuchome** [comment]
@jdsn it works with your latest patch regarding nodescompute value

**Comment by vuntz** [comment]
+1

**Comment by jsuchome** [comment]
This is required: https://github.com/SUSE-Cloud/automation/pull/327

**Comment by jdsn** [comment]
Depending PR https://github.com/SUSE-Cloud/automation/pull/327 was just merged.

**Comment by jdsn** [comment]
sorry

**Comment by MaximilianMeister** [comment]
:+1:

File: scripts/qa_crowbarsetup.sh [hunk 19]
```diff
@@ -1736,6 +1743,18 @@ function custom_configuration()
                 novanodes="[ ${novanodes%,} ]"
                 proposal_set_value nova default "['deployment']['nova']['elements']['nova-multi-compute-${libvirt_type}']" "$novanodes"
             fi
+
+            if [ -n "$want_sles12" ] && [ -n "$want_docker" ] ; then
+                proposal_set_value nova default "['deployment']['nova']['elements']['nova-multi-compute-docker']" "['$sle12node']"
+                # do not assign another compute role to this node
+                if [ -n "$nodescompute" ] ; then
```
Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: no else case here? what happens if $nodescompute is empty?

The chances are low that it will, but then we should exit here (via complain), right?

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Why? Than I assume there's just that SLE12 node assigned to nova-multi-compute-docker 3lines above

Comment (jsuchome) [inline scripts/qa_crowbarsetup.sh]: Hm, actually not, that one should be also included in nodescompute... I think if we really have no compute node, than we will have a deployment without compute node, that doesn't need to yield error, or at least not here....

Comment (jdsn) [inline scripts/qa_crowbarsetup.sh]: Ok, reading it again I think its fine though. Having no compute nodes is a different kind of bug, that should be caught earlier, not here, I agree.
